src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -759,6 +760,34 @@ PurgeConnectionForPlacement(ShardPlacement *placement)
 
 
 /*
+ * Simply removes a given connection from the transaction participant hash.
+ */
+static void
+RemoveXactConnection(PGconn *connection)
+{
+	NodeConnectionKey nodeKey;
+	NodeConnectionEntry *participantEntry = NULL;
+	bool entryFound = false;
+
+	if (xactParticipantHash == NULL)
+	{
+		return;
+	}
+
+	BuildKeyForConnection(connection, &nodeKey);
+
+	/* the participant hash doesn't use the user field */
+	MemSet(&nodeKey.nodeUser, 0, sizeof(nodeKey.nodeUser));
+	participantEntry = hash_search(xactParticipantHash, &nodeKey, HASH_FIND,
+								   &entryFound);
+
+	Assert(entryFound);
```

Comment:
I'm slightly inclined to make this an ereport(ERROR) instead. Don't want to suddenly die because of null pointer dereferences because we forgot some path (or abort() out).